docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+# Upgrade Process
+
+## Abstract
+The goal of this document is to design and describe a system where upgrading a project from one version of a Package (say Angular2 `2.X`) to an ulterior version of the same package (say Angular `2.Y`, where `Y` is strictly greater than `X`), implying refactoring the code base of said project to compile and execute properly after the upgrade.
+
+The upgrade must be transitive. If the process can be applied to go from version `X` to version `X+1`, and from `X+1` to `X+2`, then the user must be able to run the same process for going from `X` to `X+2` directly, without failure.
+
+This is done by a series of steps, called Upgrade steps, that will be applied incrementally to the project. These steps are free form functions; they take a source tree and outputs a potentially different source tree, or indicates to the user what tasks are to be performed, in case such tasks cannot be done automatically.
```

Comment:
Upgrade steps should also report what they did at the end, even if it's a basic diff output. Perhaps writing it to a upgrade log, which would be quite useful in the case of several sequential steps.